### PR TITLE
[SPARK-36068][BUILD][TEST] No tests in hadoop-cloud run unless hadoop-3.2 profile is activated explicitly

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes an issue that no tests in `hadoop-cloud` are compiled and run unless `hadoop-3.2` profile is activated explicitly.
The root cause seems similar to SPARK-36067 (#33276) so the solution is to activate `hadoop-3.2` profile in `hadoop-cloud/pom.xml` by default.

This PR introduced an empty profile for `hadoop-2.7`. Without this, building with `hadoop-2.7` fails.

### Why are the changes needed?

`hadoop-3.2` profile should be activated by default so tests in `hadoop-cloud` also should be compiled and run without activating `hadoop-3.2` profile explicitly.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Confirmed tests in `hadoop-cloud` ran with both SBT and Maven.
```
build/sbt -Phadoop-cloud "hadoop-cloud/test"
...
[info] CommitterBindingSuite:
[info] - BindingParquetOutputCommitter binds to the inner committer (258 milliseconds)
[info] - committer protocol can be serialized and deserialized (11 milliseconds)
[info] - local filesystem instantiation (3 milliseconds)
[info] - reject dynamic partitioning (1 millisecond)
[info] Run completed in 1 second, 234 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.

build/mvn -Phadoop-cloud -pl hadoop-cloud test
...
CommitterBindingSuite:
- BindingParquetOutputCommitter binds to the inner committer
- committer protocol can be serialized and deserialized
- local filesystem instantiation
- reject dynamic partitioning
Run completed in 560 milliseconds.
Total number of tests run: 4
Suites: completed 2, aborted 0
Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

I also confirmed building with `-Phadoop-2.7` successfully finishes with both SBT and Maven.
```
build/sbt -Phadoop-cloud -Phadoop-2.7 "hadoop-cloud/Test/compile"
build/mvn -Phadoop-cloud -Phadoop-2.7 -pl hadoop-cloud testCompile
```

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @HyukjinKwon @sunchao 
BTW, this kind of issue was not detected because `hadoop-3.2` is explicitly activated on CI.
Should we leave it as it is or better to remove `-Phadoop-3.2`?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I think its fine to just leave it as is for now

**Comment by sarutak** [comment]
O.K, I'll leave it as is.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.2.

**Comment by HyukjinKwon** [comment]
Oops, I rushed to merge. It seems like failing with Hadoop 2.7:

```
[error] sbt.librarymanagement.ResolveException: Error downloading org.apache.hadoop:hadoop-cloud-storage:2.7.4
[error]   Not found
[error]   Not found
[error]   not found: https://maven-central.storage-download.googleapis.com/maven2/org/apache/hadoop/hadoop-cloud-storage/2.7.4/hadoop-cloud-storage-2.7.4.pom
[error]   not found: https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-cloud-storage/2.7.4/hadoop-cloud-storage-2.7.4.pom
[error]   not found: /home/runner/.m2/repository/org/apache/hadoop/hadoop-cloud-storage/2.7.4/hadoop-cloud-storage-2.7.4.pom
[error]   not found: /home/runner/.ivy2/localorg.apache.hadoop/hadoop-cloud-storage/2.7.4/ivys/ivy.xml
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:258)
[error] 	at lmcoursier.CoursierDependencyResolution.$anonfun$update$38(CoursierDependencyResolution.scala:227)
[error] 	at scala.util.Either$LeftProjection.map(Either.scala:573)
[error] 	at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:227)
[error] 	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)
[error] 	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:59)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:133)
[error] 	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:146)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:146)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:127)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:160)
[error] 	at sbt.Classpaths$.$anonfun$updateTask0$1(Defaults.scala:3678)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] (hadoop-cloud / update) sbt.librarymanagement.ResolveException: Error downloading org.apache.hadoop:hadoop-cloud-storage:2.7.4
```

Let me revert it for now.

File: hadoop-cloud/pom.xml [hunk 1]
```diff
@@ -208,6 +208,9 @@
     -->
     <profile>
       <id>hadoop-3.2</id>
```
Comment (HyukjinKwon) [inline hadoop-cloud/pom.xml:210]: I think we should have an empty profile `hadoop-2.7` to deactivate `hadoop-3.2` when `hadoop-2.7` is set?

**Comment by sarutak** [comment]
@HyukjinKwon 
Ah, O.K. I'll fix it. Do you revert this change first?

**Comment by sarutak** [comment]
> @HyukjinKwon
Ah, O.K. I'll fix it. Do you revert this change first?

Oh, you've already done.

**Comment by HyukjinKwon** [comment]
Thank you @sarutak! sorry I rushed merging it.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by steveloughran** [comment]
* I don't think any of the tests do anything useful until that 3.2 profile is in, so the fact they don't run isn't necessarily that important.
* is it safe to be mixing default profile versions in different modules? I can see that breaking a lot of assumptions. My real fear here is that somehow you get a build where different modules are compiled against different dependencies.

**Comment by sarutak** [comment]
> I don't think any of the tests do anything useful until that 3.2 profile is in, so the fact they don't run isn't necessarily that important.

hadoop-3.2 profile is expected to be activated by default unless `hadoop-2.7` is activated explicitly right?
So, I think tests in `hadoop-cloud` are also expected to run regardless of `-Phadoop-3.2` explicitly specified or not.

> is it safe to be mixing default profile versions in different modules? I can see that breaking a lot of assumptions. My real fear here is that somehow you get a build where different modules are compiled against different dependencies.

I think, default profile in a pom affects only within the pom. So a default profile doesn't affect other modules right?

File: hadoop-cloud/pom.xml [hunk 2]
```diff
@@ -201,13 +201,22 @@
...
+      <id>hadoop-2.7</id>
+    </profile>
     <!--
      Hadoop 3 simplifies the classpath, and adds a new committer base class which
      enables store-specific committers.
     -->
     <profile>
       <id>hadoop-3.2</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
```
Comment (srowen) [inline hadoop-cloud/pom.xml:218]: BTW this has no effect if _any_ other profiles are activated manually, IIRC. Does that pose a problem?

Comment (sarutak) [reply hadoop-cloud/pom.xml:218]: I think it's true within each pom.
https://maven.apache.org/ref/3.3.9/maven-model/apidocs/org/apache/maven/model/Activation.html#setActiveByDefault(boolean)

In fact, tests run with `build/sbt -Phadoop-cloud -Pyarn -Phive -Phive-thriftserver -Pkubernetes "testOnly org.apache.spark.internal.io.cloud.*"`

**Comment by HyukjinKwon** [comment]
Hm, I think we already have some places that assumes Hadoop 3 by default (cc @dongjoon-hyun too). So making Hadoop 3 as a default is correct in any event (?). If it doesn't work somewhere unexpectedly, we would have to fix it there to comply this PR.

**Comment by sarutak** [comment]
Finally, how should we treat this issue?
One option can be leaving it as is.

**Comment by HyukjinKwon** [comment]
I think this PR looks good to me. Let me merge this in in few days if there are no more commnets on this.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by sunchao — APPROVED** [review]
+1 (non-binding) too - I think it's better to keep the modules consistent as we do this in other places.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.2.

**Pull request closed without merging**